xde/src/xde.rs

```diff
@@ -1810,78 +1892,755 @@ fn guest_loopback(
+                    + (ctx.non_eth_payl_bytes as usize)
+                    + ctx.inner_eth_len;
+                let Ok(my_pkt) =
+                    ctx.out_pkt.pullup(NonZeroUsize::new(pullup_len))
+                else {
+                    opte::engine::dbg!(
+                        "mcast TX external pullup failed: requested {} bytes",
+                        pullup_len
+                    );
+                    let xde = get_xde_state();
+                    xde.stats.vals.mcast_tx_pullup_fail().incr(1);
+                    continue;
+                };
+                match entry_state.get_by_key(*el) {
+                    Some(dev) => {
+                        // DTrace probe: local delivery
+                        let (af, addr_ptr) = match &ctx.inner_dst {
+                            oxide_vpc::api::IpAddr::Ip4(v4) => (
+                                2usize,
+                                AsRef::<[u8]>::as_ref(v4).as_ptr() as uintptr_t,
+                            ),
+                            oxide_vpc::api::IpAddr::Ip6(v6) => (
+                                26usize,
+                                AsRef::<[u8]>::as_ref(v6).as_ptr() as uintptr_t,
+                            ),
+                        };
+                        __dtrace_probe_mcast__local__delivery(
+                            af,
+                            addr_ptr,
+                            ctx.vni.as_u32() as uintptr_t,
+                            dev.port.name_cstr().as_ptr() as uintptr_t,
+                        );
+                        guest_loopback(src_dev, dev, *el, my_pkt, postbox);
+                        let xde = get_xde_state();
+                        xde.stats.vals.mcast_tx_external().incr(1);
+                    }
+                    None => {
+                        let xde = get_xde_state();
+                        xde.stats.vals.mcast_tx_stale_external().incr(1);
                     }
                 }
+            }
+        }
+    }
 
-                Ok(ProcessResult::Drop { reason }) => {
-                    opte::engine::dbg!("loopback rx drop: {:?}", reason);
+    // Underlay/infrastructure forwarding only if the merged delivery mode
+    // calls for it. External-only means local delivery only, no underlay fanout.
+    let do_underlay = matches!(
+        delivery_mode,
+        oxide_vpc::api::Replication::Underlay
+            | oxide_vpc::api::Replication::All
+    );
+
+    if do_underlay {
+        // Re-acquire locks for underlay forwarding
+        let xde = get_xde_state();
+        let mgmt = xde.management_lock.lock();
+        let mcast_fwd = mgmt.mcast_fwd.read();
+
+        if let Some(next_hops) = mcast_fwd.get(&ctx.inner_dst) {
+            // We found forwarding entries, replicate to each next hop
+            for (next_hop, replication) in next_hops.iter() {
+                // Clone packet with headers using pullup
+                let pullup_len = (ctx.encap_len as usize)
```

Comment:
All of these offsets will be identical to the above cases, nor do they need to be recomputed in the `for` loop. I don't think it's unreasonable to just unconditionally compute `pullup_len`, `geneve_offset` and friends considering they must be used at least once somewhere if we've found a valid M2P entry etc and successfully processed the packet.